File: JavaScript/src/refactor-me/ProductDataConsolidator.js
function ProductDataConsolidator() {}

const CURRENCIES_AVAILABLE = {
  NZD: {
    value: 1,
  },
  EURO: {
    value: 0.67,
  },
  USD: {
    value: 0.76,
  },
};

const LAWN_MOWER = {
	repository: new LawnmowerRepository(),
	type: "Lawnmower"
}

const PHONE_CASE = {
	repository: new PhoneCaseRepository(),
	type: "Phone Case"
}

const TSHIRT = {
	repository: new TShirtRepository(),
	type: "T-Shirt"
}

function extractProductData(repository, type, currencyFactor) {
  const products = [];

  const items = repository.getAll();
  for (let i = 0; i < items.length; i++) {
    products.push({
      id: items[i].id,
      name: items[i].name,
      price: (items[i].price * currencyFactor).toFixed(2),
      type,
    });
  }

  return products;
}

function getProductDataForCurrency(currency) {
  const products = [];

  products.push(...extractProductData(LAWN_MOWER.repository, LAWN_MOWER.type, currency.value));
  products.push(...extractProductData(PHONE_CASE.repository, PHONE_CASE.type, currency.value));
  products.push(...extractProductData(TSHIRT.repository, TSHIRT.type, currency.value));

  return products;
}

ProductDataConsolidator.get = function () {
  return getProductDataForCurrency(CURRENCIES_AVAILABLE.NZD);
};

ProductDataConsolidator.getInUSDollars = function () {
  return getProductDataForCurrency(CURRENCIES_AVAILABLE.USD);
};

ProductDataConsolidator.getInEuros = function () {
  return getProductDataForCurrency(CURRENCIES_AVAILABLE.EURO);
};

File: C#/RefactorMe/ProductDataConsolidator.cs
﻿using RefactorMe.DontRefactor.Data.Implementation;
using RefactorMe.DontRefactor.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace RefactorMe
{
    public class ProductDataConsolidator
    {
        private const double NZDExchangeRate = 1.0;
        private const double EUROExchangeRate = 0.67;
        private const double USDExchangeRate = 0.76;

        private static readonly LawnmowerRepository LawnmowerRepository = new LawnmowerRepository();
        private static readonly PhoneCaseRepository PhoneCaseRepository = new PhoneCaseRepository();
        private static readonly TShirtRepository TShirtRepository = new TShirtRepository();

        public static List<Product> Get()
        {
            var lawnmowers = LawnmowerRepository.GetAll();
            var phoneCases = PhoneCaseRepository.GetAll();
            var tShirts = TShirtRepository.GetAll();

            var products = new List<Product>();

            products.AddRange(GenerateLawnowersProducts(lawnmowers, NZDExchangeRate, "Lawnmower"));
            products.AddRange(GeneratePhoneCasesProducts(phoneCases, NZDExchangeRate, "Phone Case"));
            products.AddRange(GenerateTShirtsProducts(tShirts, NZDExchangeRate, "T-Shirt"));

            return products;
        }

        public static List<Product> GetInUSDollars()
        {
            var lawnmowers = LawnmowerRepository.GetAll();
            var phoneCases = PhoneCaseRepository.GetAll();
            var tShirts = TShirtRepository.GetAll();

            var products = new List<Product>();

            products.AddRange(GenerateLawnowersProducts(lawnmowers, USDExchangeRate, "Lawnmower"));
            products.AddRange(GeneratePhoneCasesProducts(phoneCases, USDExchangeRate, "Phone Case"));
            products.AddRange(GenerateTShirtsProducts(tShirts, USDExchangeRate, "T-Shirt"));

            return products;
        }

        public static List<Product> GetInEuros()
        {
            var lawnmowers = LawnmowerRepository.GetAll();
            var phoneCases = PhoneCaseRepository.GetAll();
            var tShirts = TShirtRepository.GetAll();

            var products = new List<Product>();

            products.AddRange(GenerateLawnowersProducts(lawnmowers, EUROExchangeRate, "Lawnmower"));
            products.AddRange(GeneratePhoneCasesProducts(phoneCases, EUROExchangeRate, "Phone Case"));
            products.AddRange(GenerateTShirtsProducts(tShirts, EUROExchangeRate, "T-Shirt"));

            return products;
        }

        private static List<Product> GenerateLawnowersProducts(IEnumerable<Lawnmower> lawnmowers, double priceFactor, string type)
        {
            return lawnmowers
                .Select(item => new Product
                {
                    Id = item.Id,
                    Name = item.Name,
                    Price = Math.Round(item.Price * priceFactor, 2),
                    Type = type
                })
                .ToList();
        }

        private static List<Product> GeneratePhoneCasesProducts(IEnumerable<PhoneCase> phoneCases, double priceFactor, string type)
        {
            return phoneCases
                .Select(item => new Product
                {
                    Id = item.Id,
                    Name = item.Name,
                    Price = Math.Round(item.Price * priceFactor, 2),
                    Type = type
                })
                .ToList();
        }

        private static List<Product> GenerateTShirtsProducts(IEnumerable<TShirt> tShirts, double priceFactor, string type)
        {
            return tShirts
                .Select(item => new Product
                {
                    Id = item.Id,
                    Name = item.Name,
                    Price = Math.Round(item.Price * priceFactor, 2),
                    Type = type
                })
                .ToList();
        }
    }
}
